### ISSUE TITLE
kubelet on the master node in a GCE cluster does not perform housekeeping tasks

### ISSUE THREAD

**Issue by yujuhong** [kubernetes/kubernetes#15122]
We configure salt the populate the http manifest source for the master node [here](https://github.com/kubernetes/kubernetes/blob/master/cluster/saltbase/salt/kubelet/default#L54). 
```
{% set manifest_url = "" -%}
{% if grains['roles'][0] == 'kubernetes-master' and grains.cloud in ['gce'] -%}
  {% set manifest_url = "--manifest-url=http://metadata.google.internal/computeMetadata/v1/instance/attributes/google-container-manifest --manifest-url-header=Metadata-Flavor:Google" -%}
{% endif -%}
```

If user doesn't supply such a file, kubelet would think not all sources are ready and would refuse to delete pods and/or performing housekeeping tasks.

A cluster created by kube-up.sh doesn't have `google-container-manifest` attribute set.
```console
$ curl "http://metadata.google.internal/computeMetadata/v1/instance/attributes/" -H "Metadata-Flavor: Google"
kube-env
startup-script
```
What periodically appears in kubelet.log if you increase the logging level to 6.
```
I1005 22:28:17.276930    4035 config.go:99] Looking for [file http], have seen map[file:{}]
```

This issue is not urgent because we don't schedule any user pods on the master node, however, it might be important for the longevity of the master node. It will need to be resolve to handle mirror pod deletions, etc #13102. One solution is to supply an empty manifest file upon cluster creation.

**Comment by dchen1107** [comment]
cc/ @roberthbailey @zmerlynn

**Comment by roberthbailey** [comment]
/cc @a-robinson 

If you do specify a manifest file in the metadata server, then you are saying that we wouldn't expect to experience the issue about skipping housekeeping. Interesting. 

I've been wondering if we should enable fetching from the metadata server on the kubelet (as a work-around to get pod-per-node on GCE before the daemon controller is ready) but this implies that that might not be the greatest idea.

**Comment by yujuhong** [comment]
This is a safety mechanism for kubelet restarts since we don't have checkpointing yet. Without this condition, kubelet may kill all running pods before seeing them from the sources.

Perhaps we can disable this flag for master in the salt configuration? Users who wants to use this feature can turn it on and create a meaningful manifest file.

**Comment by dchen1107** [comment]
@roberthbailey We should disable --manifest-url for now to make all nodes are same regardless master or minion. Today Kubelet uses this to determine if source, especially on kube-apiserver is ready, and what is desire_state of running pods on the node.

For a long run, we should introduce a local persistent store, could be etcd, or simply checkpoint file, so that node availability, or source readiness isn't coupled with apiserver availability. Once we are done with that, all nodes can fetch from metadata server on the kubelet.

**Comment by yujuhong** [comment]
> For a long run, we should introduce a local persistent store, could be etcd, or simply checkpoint file, so that node availability, or source readiness isn't coupled with apiserver availability. Once we are done with that, all nodes can fetch from metadata server on the kubelet.

+1. I think we mau need a persistent pod store earlier than we thought. This would resolve many issues. To name a few:
 - source readiness (the issue in question).
 - pod admission problem. when kubelet restarts, it'd think all pods are new and will go through the admission process again (e.g, it may reject an already running pod if disk space is not enough).
 - generating pod status w/ the absence of pod specs; this can be resolved using labels (#15089)

**Comment by thockin** [comment]
This also breaks kube-push, so I'd say it is more important than P2.

**Comment by a-robinson** [comment]
@yujuhong - are you working on disabling the manifest-url on the master by default or should I send a pr?

**Comment by dchen1107** [comment]
What is the problem you have seen with kube-push? https://github.com/kubernetes/kubernetes/pull/11963#issuecomment-145989523 introduced this a while back, and I didn't see kube-push issue.

**Comment by yujuhong** [comment]
@a-robinson, feel free to send a PR since you'll have to change the settings in GKE

**Comment by dchen1107** [comment]
I am sending pr for this.

**Comment by thockin** [comment]
https://github.com/kubernetes/kubernetes/issues/15186

Not sure if it is the same, but I just did a kube-push and EVERYTHING on master is running 2 POD containers and not cleaning up old containers.

**Comment by thockin** [comment]
Maybe we should consider a 404 on that URL to be "seen".  The server existed and there was nothing there...  a little scary..

**Comment by a-robinson** [comment]
Considering a 404 to mean things are ok but there's nothing to run doesn't seem unreasonable

**Comment by yujuhong** [comment]
> Not sure if it is the same, but I just did a kube-push and EVERYTHING on master is running 2 POD containers and not cleaning up old containers.

Confirmed that removing the manifest_url and running kube-push again restored peace on my master node.

**Comment by yujuhong** [comment]
This is pretty bad because I thought I was running the new master components after kube-push, but that was not true. kubelet couldn't start the new containers because the old ones weren't killed.

**Comment by dchen1107** [comment]
Ok, we need to patch 1.1 release here.

**Comment by dchen1107** [comment]
#15190 was merged and cherry-pick #15398 for release 1.1. Closed the issue.
